src/drafter.cc

```diff
@@ -0,0 +1,204 @@
+// vi:cin:et:sw=4 ts=4
+//
+//  drafter.cc
+//
+//  Created by Jiri Kratochvil on 2015-02-11
+//  Copyright (c) 2015 Apiary Inc. All rights reserved.
+//
+//
+//
+
+#include <string>
+#include <iostream>
+#include <sstream>
+#include <fstream>
+
+#include <tr1/memory>
+#include <tr1/shared_ptr.h>
+
+#include "snowcrash.h"
+#include "SectionParserData.h"  // snowcrash::BlueprintParserOptions
+
+#include "sos.h"
+#include "sosJSON.h"
+#include "sosYAML.h"
+
+#include "SerializeAST.h"
+#include "SerializeSourcemap.h"
+
+#include "reporting.h"
+#include "config.h"
+
+namespace sc = snowcrash;
+
+/**
+ *  \brief deleter functor to shared_ptr<> which does nothing
+ *  usefull to get system resource like std::cin, std::cout to shared_ptr<>
+ */
+template<typename T>
+struct dummy_deleter {
+    void operator()(T* obj) const 
+    {
+      // do nothing
+    }
+};
+
+/**
+ *  \brief functor returns cin/cout ptr in based on istream or ostream
+ */
+template<typename T> struct std_io;
+
+template<> struct std_io<std::istream> {
+    std::istream* operator()() const
+    { 
+        return &std::cin; 
+    }
+};
+
+template<> struct std_io<std::ostream> {
+    std::ostream* operator()() const
+    { 
+        return &std::cout; 
+    }
+};
+
+/**
+ *  \brief functor returns appropriate cin/cout based on istream/ostream
+ */
+template<typename T> struct std_io_selector {
+    typedef T stream_type;
+    typedef std::tr1::shared_ptr<stream_type> return_type;
+
+    return_type operator()() const
+    { 
+        return return_type(std_io<T>()(), dummy_deleter<stream_type>()); 
+    }
+};
+
+/**
+ *  \brief functor select fstream based on iostream
+ */
+template <typename Stream> struct to_fstream;
```

Comment:
Yes, it is intentionally, to hint paremeter must be iostream, but you are right it differs from rest of file whewre is used <T>. 
Will fix